### PR TITLE
java doc

### PR DESCRIPTION
**Pull request by max402**

### CODE REVIEW

File: api/src/main/java/de/adorsys/keymanagement/api/generator/EncryptingKeyGenerator.java [hunk 1]
```diff
@@ -3,7 +3,15 @@
...
 import de.adorsys.keymanagement.api.types.template.generated.Encrypting;
 
+/**
+ * Specific interface for generating encryption keys
+ */
 public interface EncryptingKeyGenerator {
 
+    /**
+     * generates key pair using template
+     * @param fromTemplate encrypting template which is used to generate key pair
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/generator/EncryptingKeyGenerator.java]: It is better to expand `encrypting template which is used to generate key pair` to `Template that defines encrypting key properties like encryption algorithm, key size, key name that are going to be used on creation`

File: api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java [hunk 2]
```diff
@@ -4,9 +4,29 @@
...
 import de.adorsys.keymanagement.api.types.template.provided.ProvidedKey;
 
+/**
+ * Specific interface for generating secret keys
+ */
 public interface SecretKeyGenerator {
 
+    /**
+     * Generates key from template
+     * @param fromTemplate password based encryption entry template
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java]: Expand fromTemplate definition like above.

File: api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java [hunk 3]
```diff
@@ -4,9 +4,29 @@
...
+     * @return generated key
+     */
     ProvidedKey generate(Pbe fromTemplate);
+
+    /**
+     * Generates key from template using default salt and iterations count
+     * @param fromTemplate password based encryption entry template
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java]: Expand fromTemplate definition like above.

File: api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java [hunk 4]
```diff
@@ -4,9 +4,29 @@
...
+     * @return generated key
+     */
     ProvidedKey generateRaw(Pbe fromTemplate);
+
+    /**
+     * Generates key from Secret template
+     * @param fromTemplate Secret template
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java]: Expand fromTemplate definition like above.

File: api/src/main/java/de/adorsys/keymanagement/api/generator/SigningKeyGenerator.java [hunk 5]
```diff
@@ -3,7 +3,15 @@
...
 import de.adorsys.keymanagement.api.types.template.generated.Signing;
 
+/**
+ * Specific interface for generating signing keys
+ */
 public interface SigningKeyGenerator {
 
+    /**
+     * generates key pair using template
+     * @param fromTemplate encrypting template which is used to generate key pair
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/generator/SigningKeyGenerator.java]: Expand fromTemplate definition like above.

File: api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreOper.java [hunk 6]
```diff
@@ -8,5 +8,12 @@
 
 public interface KeyStoreOper extends KeyStoreCreator {
 
+    /**
+     * Adds key entry to keystore.
+     * @param ks keystore where to add key
+     * @param entry key entry which will be added to keystore
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreOper.java]: `Key that is going to be added to KeyStore. Note that ProvidedKeyTemplate is the wrapper for KeyStore entries that provides additional information like key name`

File: api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java [hunk 7]
```diff
@@ -35,7 +35,23 @@
      */
     <Q> AliasView<Q> allAliases();
 
+    /**
+     * Get source of keys
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java]: Get source of keys. For example KeySource can wrap following entities - java Keystore, Database table, list of Key instances, etc.

File: api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java [hunk 8]
```diff
@@ -35,7 +35,23 @@
...
+     * @return object which have access to keys and can operate with them
+     */
     KeySource source();
+
+    /**
+     * Generates keySet from all existing in key source keys
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java]: `Clones current KeySource into KeySet in a way, that all keys will be materialized in KeySet. You can view it as generates keySet from all existing in key source keys`

File: api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java [hunk 9]
```diff
@@ -35,7 +35,23 @@
...
+     * @param keyPassword password which is used to access keys
+     * @return keySet with lists of key pairs, secret keys or key entries
+     */
     KeySet copyToKeySet(Function<String, char[]> keyPassword);
+
+    /**
+     * Generates template from all existing in key source keys
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java]: `Clones current KeySource into KeySetTemplate in a way, that all keys will be materialized in KeySetTemplate. You can view it as generating template from all existing in key source keys`

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java [hunk 10]
```diff
@@ -8,15 +8,52 @@
 
 public interface KeyMetadataOper {
 
+    /**
+     * By default password is null for keys metadata
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: `Password that will be used to store metadata as key entry in encrypted KeySource (i.e. java Keystore).  By default password is null for keys of metadata type`

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java [hunk 11]
```diff
@@ -8,15 +8,52 @@
...
+     * @return password for specified alias as byte array
+     */
     @Nullable
     @SuppressFBWarnings("PZLA_PREFER_ZERO_LENGTH_ARRAYS")
     default char[] metadataPassword(String forAlias) {
         return null;
     }
 
+    /**
+     * Generates metadata alias from key alias
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: Generates metadata key entry alias from key alias

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java [hunk 12]
```diff
@@ -8,15 +8,52 @@
...
+     * @param forAlias key name used to find key in keystore
+     * @return metadata alias
+     */
     String metadataAliasForKeyAlias(String forAlias);
+
+    /**
+     * Checks if entry contains metadata
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: Checks if key for provided entry alias is metadata-entry

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java [hunk 13]
```diff
@@ -8,15 +8,52 @@
...
+     * @param forAlias name used to find key
+     * @param keyStore Keystore where search entry
+     * @return true if entry contains metadata
+     */
     boolean isMetadataEntry(String forAlias, KeyStore keyStore);
+
+    /**
+     * Reads metadata from specified by alias key found in specified keyStore
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: Reads metadata entry for key (by its alias) from KeyStore

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java [hunk 14]
```diff
@@ -8,15 +8,52 @@
...
+     * @param forAlias name used to find key
+     * @param keyStore Keystore where search key
+     * @return key metadata
+     */
     KeyMetadata extract(String forAlias, KeyStore keyStore);
+
+    /**
+     * Saves metadata to existing key in keystore
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: Saves metadata for existing key in KeyStore

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java [hunk 15]
```diff
@@ -8,15 +8,52 @@
...
+     * @param forAlias name used to find key
+     * @param metadata new key metadata
+     * @param keyStore which consists key to which metadata will be added
+     */
     void persistMetadata(String forAlias, KeyMetadata metadata, KeyStore keyStore);
+
+    /**
+     * Removes metadata from key found by specified alias from specified keystore
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: Removes key metadata from KeyStore

File: api/src/main/java/de/adorsys/keymanagement/api/source/KeySource.java [hunk 16]
```diff
@@ -12,18 +12,65 @@
...
 
+    /**
+     * Removes key by keyId from key source
+     * @param keyId key alias
+     */
     void remove(String keyId);
+
+    /**
+     * Adds key to key source
+     * @param keyTemplate keeps key, key metadata and generates new alias for the key
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/source/KeySource.java]: keyTemplate keeps key, key metadata and acts as alias source (either generate alias or hardcode it) for the key

File: api/src/main/java/de/adorsys/keymanagement/api/types/template/GeneratedKeyTemplate.java [hunk 17]
```diff
@@ -4,5 +4,9 @@
 
 public interface GeneratedKeyTemplate extends KeyTemplate {
 
+    /**
+     * Access to metadata stored in key template
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/types/template/GeneratedKeyTemplate.java]: Gets metadata associated with KeyTemplate

File: api/src/main/java/de/adorsys/keymanagement/api/types/template/KeyNamingStrategy.java [hunk 18]
```diff
@@ -1,6 +1,12 @@
 package de.adorsys.keymanagement.api.types.template;
 
+/**
+ * Interface which determines algorithm for generating key name
+ */
 public interface KeyNamingStrategy {
 
+    /**
+     * @return generated key name
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/types/template/KeyNamingStrategy.java]: generated or hardcoded key name

File: api/src/main/java/de/adorsys/keymanagement/api/types/template/KeyTemplate.java [hunk 19]
```diff
@@ -4,6 +4,15 @@
 
 public interface KeyTemplate {
 
+    /**
+     * Generates alias for new key
+     * @return generated key name
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/types/template/KeyTemplate.java]: generated or hardcoded key name

File: api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java (hunk 4 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java (hunk 3 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java (hunk 2 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/generator/SecretKeyGenerator.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/generator/EncryptingKeyGenerator.java (hunk 1 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/generator/EncryptingKeyGenerator.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/generator/SigningKeyGenerator.java (hunk 5 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/generator/SigningKeyGenerator.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreOper.java (hunk 6 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreOper.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/types/template/KeyTemplate.java (hunk 19 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/types/template/KeyTemplate.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/types/template/KeyNamingStrategy.java (hunk 18 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/types/template/KeyNamingStrategy.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/types/template/GeneratedKeyTemplate.java (hunk 17 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/types/template/GeneratedKeyTemplate.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/source/KeySource.java (hunk 16 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/source/KeySource.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java (hunk 15 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java (hunk 14 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java (hunk 13 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java (hunk 12 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java (hunk 11 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java (hunk 10 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/metadata/KeyMetadataOper.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java (hunk 9 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java (hunk 8 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java (hunk 7 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/keystore/KeyStoreView.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/view/View.java [hunk 20]
```diff
@@ -14,7 +19,23 @@
      */
     QueryResult<O> retrieve(String query);
 
+    /**
+     * Guarantees result uniqueness
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/view/View.java]: Returns exactly one result, if ResultSet has zero or more throws an Exception

File: api/src/main/java/de/adorsys/keymanagement/api/view/View.java [hunk 21]
```diff
@@ -14,7 +19,23 @@
...
+     * @param query understandable to implementation query.
+     * @return query result
+     */
     O uniqueResult(Q query);
+
+    /**
+     * Guarantees result uniqueness
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/view/View.java]: Returns exactly one result, if ResultSet has zero or more throws an Exception

File: api/src/main/java/de/adorsys/keymanagement/api/view/View.java [hunk 22]
```diff
@@ -14,7 +19,23 @@
...
+     * @param query understandable to implementation query as a string
+     * @return query result
+     */
     O uniqueResult(String query);
+
+    /**
+     * Retrieves all elements of target object
```
Comment (valb3r) [inline api/src/main/java/de/adorsys/keymanagement/api/view/View.java]: Retrieves all elements (like aliases for `{@link ...AliasView}`) of target object

File: api/src/main/java/de/adorsys/keymanagement/api/view/View.java (hunk 20 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/view/View.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/view/View.java (hunk 21 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/view/View.java]: done

File: api/src/main/java/de/adorsys/keymanagement/api/view/View.java (hunk 22 above)
Comment (max402) [reply api/src/main/java/de/adorsys/keymanagement/api/view/View.java]: done

File: .travis/push-javadoc-to-gh-pages.sh [hunk 23]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+if [[ "$TRAVIS_REPO_SLUG" == "adorsys/keystore-management"
```
Comment (valb3r) [inline .travis/push-javadoc-to-gh-pages.sh]: Why not to return immediately if not matches

File: .travis/push-javadoc-to-gh-pages.sh [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+    && "$TRAVIS_JDK_VERSION" == "openjdk8"
+    && "$TRAVIS_PULL_REQUEST" == "false"
+    &&  "$TRAVIS_TAG" ]];
+then
+
+  mvn clean javadoc:aggregate -P javadoc
+
+  echo -e "Publishing javadoc...\n"
+
+  git clone --quiet --branch=gh-pages https://"$GITHUB_TOKEN"@github.com/"$TRAVIS_REPO_SLUG" gh-pages >/dev/null
```
Comment (valb3r) [inline .travis/push-javadoc-to-gh-pages.sh]: Add space after '>'
`gh-pages > /dev/null `

File: .travis/deploy.sh [hunk 25]
```diff
@@ -6,3 +6,5 @@ echo "$GPG_OWNERTRUST" | base64 --decode | "$GPG_EXECUTABLE" --import-ownertrust
 set -e
 
 mvn --settings .travis/settings.xml package gpg:sign deploy -Prelease -DskipTests -B -U;
+
+bash .travis/push-javadoc-to-gh-pages.sh
```
Comment (valb3r) [inline .travis/deploy.sh]: Maybe it is better to match if publishing needed directly from travis.yml

File: .travis/push-javadoc-to-gh-pages.sh [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+    cd javadoc || exit
+    rm latest
+    ln -s "$TRAVIS_TAG" latest
+  )
+
+  git config --global user.email "travis@travis-ci.org"
+  git config --global user.name "travis-ci"
+  git add -f .
+  git commit -m "Latest javadoc on successful travis build $TRAVIS_BUILD_NUMBER for tag $TRAVIS_TAG auto-pushed to gh-pages"
+  git push -fq origin gh-pages >/dev/null
```
Comment (valb3r) [inline .travis/push-javadoc-to-gh-pages.sh]: Add space after '>'

File: .travis/push-javadoc-to-gh-pages.sh (hunk 24 above)
Comment (max402) [reply .travis/push-javadoc-to-gh-pages.sh]: done

File: .travis/push-javadoc-to-gh-pages.sh (hunk 26 above)
Comment (max402) [reply .travis/push-javadoc-to-gh-pages.sh]: done

File: .travis/deploy.sh (hunk 25 above)
Comment (max402) [reply .travis/deploy.sh]: I did it first as a separate deploy provider, but I don't see any cases when publishing javadoc will be untethered from release and drawback of additional deploy step is longer build

File: .travis/push-javadoc-to-gh-pages.sh (hunk 23 above)
Comment (max402) [reply .travis/push-javadoc-to-gh-pages.sh]: done

**Pull request merged**